core/envelope-sim/src/main/kotlin/fr/sncf/osrd/envelope_sim/etcs/ETCSBrakingSimulator.kt

```diff
@@ -239,22 +253,104 @@ class ETCSBrakingSimulatorImpl(override val context: EnvelopeSimContext) : ETCSB
             }
             if (etcsRanges.contains(stopOffset.distance)) {
-                endsOfAuthority.add(
-                    EndOfAuthority(
-                        offsetEOA = stopOffset,
-                        // On a closed signal, we follow the indication speed curve with an SVL at
-                        // the next danger point to protect
-                        // On an open signal, we follow the permitted speed curve with no SVL
-                        offsetSVL =
-                            if (stop.rjsReceptionSignal.isStopOnClosedSignal)
-                                this.context.etcsContext.getDangerPoint(stopOffset)
-                            else null,
-                        usedCurveType =
-                            if (stop.rjsReceptionSignal.isStopOnClosedSignal) IND else PS
-                    )
-                )
+                val eoa =
+                    when (eoaType) {
+                        EoaType.STOP -> computeStopEoA(stopOffset, stop.second, eoaType)
+                        EoaType.SPACING ->
+                            computeSpacingConflictEoa(
+                                stopOffset,
+                                stop.second,
+                                eoaType,
+                                envelope.endPos
+                            )
+                        EoaType.ROUTING ->
+                            computeRoutingConflictEoa(stopOffset, stop.second, eoaType)
+                    }
+                if (eoa != null) endsOfAuthority.add(eoa)
             }
         }
         return endsOfAuthority
     }
+
+    /** Compute the EoA for a simple stop. */
+    private fun computeStopEoA(
+        stopOffset: Offset<TravelledPath>,
+        isStopOnClosedSignal: Boolean,
+        eoaType: EoaType
+    ): EndOfAuthority {
+        assert(eoaType == EoaType.STOP)
+        return if (isStopOnClosedSignal) {
+            // On a closed signal, EoA = signal, SvL = next closed signal danger point.
+            EndOfAuthority(
+                offsetEOA = stopOffset,
+                offsetSVL = this.context.etcsContext!!.getMandatoryDangerPoint(stopOffset),
+                usedCurveType = IND,
+                eoaType = eoaType
+            )
+        } else {
+            // On an open signal, EoA = stop, SvL = null and the used curve is the permitted speed.
+            EndOfAuthority(
+                offsetEOA = stopOffset,
+                offsetSVL = null,
+                usedCurveType = PS,
+                eoaType = eoaType
+            )
+        }
+    }
+
+    /** Compute the EoA for a spacing conflict on a signal. */
+    private fun computeSpacingConflictEoa(
+        signalOffset: Offset<TravelledPath>,
+        isInterlocking: Boolean,
+        eoaType: EoaType,
+        endPos: Double
+    ): EndOfAuthority? {
+        assert(eoaType == EoaType.SPACING)
+        val nextDetector = this.context.etcsContext!!.getNextDetector(signalOffset)!!
```

Comment:
The `!!` on next detector seems a bit optimistic. Is there now way it could be null? Like at signaling system transitions, or near a buffer stop?